Fire track event for modal

tomalec opened the pull request:
### Changes proposed in this Pull Request:

- Support multiline Track Event properties' description.
- Fire track event for modal

implements part of #232.




### Screenshots:



![modal open](https://user-images.githubusercontent.com/17435/144649805-c83f9f19-6db5-43a7-9bee-7ad81b92330a.gif)
![modal event](https://user-images.githubusercontent.com/17435/144649808-4476c373-1fcc-44ca-90eb-b59344835689.gif)

### Detailed test instructions:

0. Enable tracks logging, by `localStorage.setItem('debug', 'wc-admin*')`
1. Go to [`/wp-admin/admin.php?page=wc-admin&path=%2Fpinterest%2Fonboarding`](https://wp581wc580.test/wp-admin/admin.php?page=wc-admin&path=%2Fpinterest%2Fonboarding) 
2. (Connect account)
3. Click on "disconnect"
	- you should get `wcadmin_pfw_modal_open: { context: 'account-disconnection' }`
4. Click on overlay | "x" | "cancel" | press <kbd>Esc</kbd>
	- you should get `wcadmin_pfw_modal_closed: { context: 'account-disconnection', action: 'dismiss' }`
5. Reopen, click "Yes, I'm sure"
	- you should get `wcadmin_pfw_modal_closed: { context: 'account-disconnection', action: 'confirm' }`

(no changelog entry, we will add a single one for all tracks)
<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

tomalec committed:
Support multiline Track Event properties' description.

tomalec committed:
Fire track event for modal,

implements part of https://github.com/woocommerce/pinterest-for-woocommerce/issues/232

eason9487 commented on assets/source/setup-guide/app/components/Account/Connection.js:
```diff
@@ -44,12 +44,29 @@ const PinterestLogo = () => {
+ * @event wcadmin_pfw_modal_closed
+ * @property {string} context Indicate which modal this is for.
+ * @property {string} action
+ * 				`confirm` - When the final "got it" button is clicked.
```
💅  Looks like it should be the "Yes, I'm sure" button.

eason9487 commented on woo-tracking-jsdoc/publish.js:
```diff
@@ -80,7 +80,12 @@ exports.publish = function ( data ) {
+					mdResult += `\`${
+						property.name
+					}\` | \`${ type }\` | ${ description.replace(
+						/\s*\n\s*/,
```
💅 
How about adding the `g` flag in case of the `description` has more than one newline char?

eason9487 reviewed (commented):
Tested locally. All events were triggered as expected.

But there's a ❓ or ❔  concern. The upper-layer component `<SetupAccount>`, which uses `<AccountConnection>`, is also used in the Connection tab: `/wp-admin/admin.php?page=wc-admin&path=%2Fpinterest%2Fconnection`.

![image](https://user-images.githubusercontent.com/17420811/145004322-fc801614-f330-4f64-accf-754b25a0d7ea.png)

Do we need to make it easier to distinguish which page the model event was triggered in the event tracking data? And just noticed the tracking events `pfw_account_disconnect_button_click` of disconnecting button have the same concern.

tomalec committed:
Fix the copy in `wcadmin_pfw_modal_closed` event docs,

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/290#discussion_r763756953

tomalec committed:
Allow multipel new-lines in event property description,

Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/290#discussion_r763778061

Co-authored-by: Eason Su <eason.su.tw@gmail.com>

tomalec commented:
Regarding 
> But there's a :question: or :grey_question: concern. The upper-layer component `<SetupAccount>`, which uses `<AccountConnection>`, is also used in the Connection tab: `/wp-admin/admin.php?page=wc-admin&path=%2Fpinterest%2Fconnection`.

I'm not 100% how to approach this
[Track Events spreadsheet](https://docs.google.com/spreadsheets/d/1u25mOcsODgcdPMy_rIagUAWHCfzsvJI1e8UPrJvYtVE/edit#gid=903716843) states: 
> I17: Context is used to indicate which modal this is for

If the question is "Which modal", one may say, that on both pages this is the same "Account Disconnection" modal, exactly same content, same action, same purpose. Just opened from a different location.

So, IMHO, we are back into `path` vs `context` discussion.

tomalec commented:
On the other hand. The event name is `modal_open`, not `disconnect_modal_open`. If it's "Q: _what modal?_, A: _Account Disconnection_", then the property should be more like a `name` or `type`.
If it's `context` it rather matches the "Q: In what context this exact modal was opened?".

Now, given in the entire Plugin, we have a single **type/name/kind** of modal - the Account Disconnection modal. Which is used on two different pages/locations/context. I'll change the event description (and implementation), to something like:
```
 * @event wcadmin_pfw_modal_open 
 * Fired when any modal is open. 
 * @property {string} name What modal is it? // To be future prepared for other kinds of modals
 * @property {string} context Indicate in what context it was opened.  // IMHO, We could use `path` instead, but let's leave it as it was for now.
```

tomalec committed:
Set different context for `pfw_modal_*` and `pfw_documentation_link_click`

Track Events, in settings and wizard respectively.
Addresses https://github.com/woocommerce/pinterest-for-woocommerce/pull/290#pullrequestreview-824948568.

tomalec commented:
@eason9487  Thanks for the review, I addressed all your comments, I believe it's ready for another round.

tomalec committed:
Add a basic test for `wcadmin_pfw_modal_open`.

tomalec committed:
Add tests for `pfw_modal_closed`,

Add `@testing-library/user-event` devDependency.

tomalec commented:
I'm not 100% sure about the tests from https://github.com/woocommerce/pinterest-for-woocommerce/pull/290/commits/be2191e77ea126295303a3544a8f1c6483bfb1ba, as that's actually testing the `<Modal>` features, maybe I should rather stub `<Modal>` and check that we fire an event whenever it's dismissed? WDYT?

eason9487 reviewed (approved):
LGTM. Thanks for the elaborate thoughts!

I also think it's more reasonable to use `name` to indicate which modal it is, and use `context` first to indicate the modal is opened for/from what process/place. About whether there're more consistent and suitable properties, would be happy to discuss later.

> I'm not 100% sure about the tests from [be2191e](https://github.com/woocommerce/pinterest-for-woocommerce/commit/be2191e77ea126295303a3544a8f1c6483bfb1ba), as that's actually testing the `<Modal>` features, maybe I should rather stub `<Modal>` and check that we fire an event whenever it's dismissed? WDYT?

I think it's ok, after all, the tracking events are tied to the modal behaviors. 😂
